File: Quasics2024Code/MaeTester2024/src/main/cpp/commands/BreathingLights.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "commands/BreathingLights.h"

// Just going to adjust green for now.  Affect others later.
BreathingLights::BreathingLights(Lights*lights)
  : m_lights(lights),
  red(0),
  green(0),
  blue(0) {
  // Use addRequirements() here to declare subsystem dependencies.
  AddRequirements(m_lights);
}

// Called when the command is initially scheduled.
void BreathingLights::Initialize() {
  m_lights->SetStripColor(0, 0, 0);
  red = 0;
  green = 0;
  blue = 0;
  isIncrementing = true;
}

// Called repeatedly when this Command is scheduled to run
void BreathingLights::Execute() {
  if (green >= 255) {
    isIncrementing = false;
  }
  if (green <= 0) {
    isIncrementing = true;
  }
  if (isIncrementing && green < 255) {
    green = green + 1;
  }
  if (!isIncrementing && green > 0) {
    green = green - 1;
  }
  m_lights->SetStripColor(0, green, 0);
}

// Called once the command ends or is interrupted.
void BreathingLights::End(bool interrupted) {
  m_lights->SetStripColor(0, 0, 0);
}

// Returns true when the command should end.
bool BreathingLights::IsFinished() {
  return false;
}

File: Quasics2026Code/MaeTester2026-Java/src/main/java/frc/robot/commands/BreathingLights.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.subsystems.Lights;

public class BreathingLights extends Command {
  final Lights m_lights;
  int m_red = 0, m_green = 0, m_blue = 0;
  boolean m_isIncrementing = true;

  /** Creates a new BreathingLights. */
  public BreathingLights(Lights lights) {
    m_lights = lights;
    addRequirements(lights);
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {
    m_lights.setStripColor(Lights.BLACK);
    m_red = 0;
    m_green = 0;
    m_blue = 0;
    m_isIncrementing = true;
  }

  // Called every time the scheduler runs while the command is scheduled.
  // CODE_REVIEW: This is ignoring red and blue; despite having data members for
  // these, we only ever "breathe green".
  @Override
  public void execute() {
    if (m_green >= 255) {
      m_isIncrementing = false;
    }
    if (m_green <= 0) {
      m_isIncrementing = true;
    }
    if (m_isIncrementing && m_green < 255) {
      m_green = m_green + 1;
    }
    if (!m_isIncrementing && m_green > 0) {
      m_green = m_green - 1;
    }
    m_lights.setStripColor(0, m_green, 0);
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {
    m_lights.setStripColor(Lights.BLACK);
  }
}
